Content.Shared/Storage/EntitySystems/BinSystem.cs

```diff
@@ -79,100 +81,130 @@ private void OnEntRemoved(Entity<BinComponent> ent, ref EntRemovedFromContainerM
             return;
 
         ent.Comp.Items.Remove(args.Entity);
+        Dirty(ent);
     }
 
-    private void OnInteractHand(EntityUid uid, BinComponent component, InteractHandEvent args)
+    private void OnInteractHand(Entity<BinComponent> entity, ref InteractHandEvent args)
     {
         if (args.Handled)
             return;
 
-        EntityUid? toGrab = component.Items.LastOrDefault();
-        if (!TryRemoveFromBin(uid, toGrab, component))
+        if (!TryRemoveFromBin(entity, out var toGrab))
             return;
 
         _hands.TryPickupAnyHand(args.User, toGrab.Value);
-        _admin.Add(LogType.Pickup, LogImpact.Low,
-            $"{ToPrettyString(uid):player} removed {ToPrettyString(toGrab.Value)} from bin {ToPrettyString(uid)}.");
         args.Handled = true;
     }
 
-    /// <summary>
-    /// Alt interact acts the same as interacting with your hands normally, but allows fallback interaction if the item
-    /// has priority. E.g. a water cup on a water cooler fills itself on a normal click,
-    /// but you can use alternative interactions to restock the cup bin
-    /// </summary>
-    private void OnAltInteractHand(EntityUid uid, BinComponent component, GetVerbsEvent<AlternativeVerb> args)
+    private void OnGetAltVerbs(Entity<BinComponent> entity, ref GetVerbsEvent<AlternativeVerb> args)
     {
-        if (args.Using != null)
+        if (!args.CanAccess || !args.CanInteract)
+            return;
+
+        var user = args.User;
+        if (args.Using != null && CanInsertIntoBin(entity, args.Using.Value))
+        {
+            var used = args.Using.Value;
+            AlternativeVerb verb = new()
+            {
+                Act = () => TryInsertIntoBin(entity, used, user),
+                Icon = entity.Comp.InsertIcon,
+                Text = Loc.GetString("place-item-verb-text", ("subject", used)),
+                Priority = 2
+            };
+            args.Verbs.Add(verb);
+        }
+        else if (TryGetLastItem(entity, out var subject))
```

Comment:
Two alternatives to consider, if you want:

```suggestion
        else if (entity.Comp.Items.LastOrDefault() is { Valid: true } subject)
```

```suggestion
        else if (entity.Comp.Items.Cast<EntityUid?>().LastOrDefault() is { } subject)
```

And then can do the same thing at the other call site so you can then just remove TryGetLastItem.